### PR TITLE
feat: update identity

### PR DESCRIPTION
**Pull request by shuplenkov**
## Issue being fixed or feature implemented

https://github.com/dashpay/dips/blob/master/dip-0011.md#identity-update-transition

## What was done?

Implemented identity update state transitions

## How Has This Been Tested?

With tests

## Breaking Changes

No

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/js-dpp/lib/errors/consensus/state/identity/IdentityPublicKeyIsReadOnlyError.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+const AbstractStateError = require('../AbstractStateError');
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/state/identity/IdentityPublicKeyIsReadOnlyError.js:1]: Must be in basic

File: packages/js-dpp/lib/errors/consensus/state/identity/IdentityNotFoundError.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+const AbstractStateError = require('../AbstractStateError');
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/state/identity/IdentityNotFoundError.js:1]: It's happening during signature validation

File: packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/internal/createIdentityUpdateTransition.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  publicKeysDisabledAt?: number,
+): Promise<any> {
+  await platform.initialize();
+
+  const { dpp } = platform;
+
+  // @ts-ignore
+  const identityUpdateTransition = dpp.identity.createIdentityUpdateTransition(
+    identity.getId(),
+    identity.getRevision() + 1,
```
Comment (shumkov) [inline packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/internal/createIdentityUpdateTransition.ts:31]: Let's increment in DPP

File: packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/internal/createIdentityUpdateTransition.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * Creates a funding transaction for the platform identity and returns one-time key to sign the state transition
+ * @param {Platform} platform
+ * @param {PrivateKey} privateKey - private key to sign
+ * @param {Identity} identity - identity to update
+ * @param {IdentityPublicKey[]} addPublicKeys - public keys to add
+ * @param {number[]} disablePublicKeys - public key IDs to disable
+ * @param {number} publicKeysDisabledAt - timestamp to disable at
+ * @return {{identity: Identity, identityUpdateTransition: IdentityUpdateTransition}}
+ */
+export default async function createIdentityUpdateTransition(
```
Comment (shumkov) [inline packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/internal/createIdentityUpdateTransition.ts:16]: If we are using signStateTransition we don't need this function anymore

File: packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/internal/createIdentityUpdateTransition.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  const { dpp } = platform;
+
+  // @ts-ignore
+  const identityUpdateTransition = dpp.identity.createIdentityUpdateTransition(
+    identity.getId(),
+    identity.getRevision() + 1,
+    addPublicKeys,
+    disablePublicKeys,
+    publicKeysDisabledAt,
```
Comment (shumkov) [inline packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/internal/createIdentityUpdateTransition.ts:34]: Set in DPP

File: packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/update.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ * @param {number} publicKeysDisabledAt - timestamp to disable at
+ *
+ * @returns {boolean}
+ */
+export async function update(
+  this: Platform,
+  identityId: Identifier | string,
+  addPublicKeys?: IdentityPublicKey[],
+  disablePublicKeys?: number[],
+  publicKeysDisabledAt?: number,
```
Comment (shumkov) [inline packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/update.ts:27]: Let's repeat interface like in the DPP factory

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    return this.addPublicKeys;
+  }
+
+  /**
+   * Set public keys to add to the Identity.
+   *
+   * @param {IdentityPublicKey[]} publicKeys
+   * @returns {IdentityUpdateTransition}
+   */
+  setAddPublicKeys(publicKeys) {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js:114]: setPublicKeysToAdd

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+  getOwnerId() {
+    return this.identityId;
+  }
+
+  /**
+   * Get public keys to add to the Identity.
+   *
+   * @returns {IdentityPublicKey[]}
+   */
+  getAddPublicKeys() {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js:104]: getPulicKeysToAdd

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    return this;
+  }
+
+  /**
+   *
+   * Get Identity Public key IDs to disable for the Identity.
+   *
+   * @returns {number[]}
+   */
+  getDisablePublicKeys() {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js:126]: getPublicKeysToDisable

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+  }
+
+  /**
+   *
+   * Set Identity Public key IDs to disable for the Identity.
+   *
+   * @param {number[]} publicKeys
+   * @returns {IdentityUpdateTransition}
+   */
+  setDisablePublicKeys(publicKeyIds) {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.js:137]: setPublicKeyIdsToDisable

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/applyIdentityUpdateTransitionFactory.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    if (stateTransition.getAddPublicKeys()) {
+      const identityPublicKeys = identity
+        .getPublicKeys()
+        .concat(stateTransition.getAddPublicKeys());
+
+      identity.setPublicKeys(identityPublicKeys);
+    }
+
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/applyIdentityUpdateTransitionFactory.js:42]: Missed `storeIdentityPublicKeyHashes`

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    const addPublicKeys = stateTransition.getAddPublicKeys();
+    if (addPublicKeys) {
+      addPublicKeys.forEach((pk) => identity.getPublicKeys().push(pk));
+    }
+
+    const disablePublicKeys = stateTransition.getDisablePublicKeys();
+
+    if (disablePublicKeys) {
+      disablePublicKeys.forEach((id) => {
+        if (identity.getPublicKeyById(id).getReadOnly()) {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:58]: Check that id is exist

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    const addPublicKeys = stateTransition.getAddPublicKeys();
+    if (addPublicKeys) {
+      addPublicKeys.forEach((pk) => identity.getPublicKeys().push(pk));
+    }
+
+    const disablePublicKeys = stateTransition.getDisablePublicKeys();
+
+    if (disablePublicKeys) {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:56]: Check disable on original data. There is no point to add and disable the same identities.

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+   * @param {IdentityUpdateTransition} stateTransition
+   * @return {Promise<ValidationResult>}
+   */
+  // eslint-disable-next-line no-unused-vars
+  async function validateIdentityUpdateTransitionState(stateTransition) {
+    const result = new ValidationResult();
+
+    const identityId = stateTransition.getIdentityId();
+    const identity = await stateRepository.fetchIdentity(identityId);
+
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:33]: I would suggest to clone Identity since you changing it (`new Identity(identity.toObject())`)

File: packages/js-dpp/lib/test/fixtures/getIdentityUpdateTransitionFixture.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    assetLockProof: getInstantAssetLockProofFixture().toObject(),
+    identityId: generateRandomIdentifier(),
+    revision: 0,
+    addPublicKeys: [
+      {
+        id: 3,
+        type: IdentityPublicKey.TYPES.ECDSA_SECP256K1,
+        data: Buffer.from('AkVuTKyF3YgKLAQlLEtaUL2HTditwGILfWUVqjzYnIgH', 'base64'),
+        purpose: IdentityPublicKey.PURPOSES.AUTHENTICATION,
+        securityLevel: IdentityPublicKey.SECURITY_LEVELS.MASTER,
```
Comment (shumkov) [inline packages/js-dpp/lib/test/fixtures/getIdentityUpdateTransitionFixture.js:21]: We shouldn't be able to send with `disabledAt`. It must be changed only programmatically.

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    if (disablePublicKeys) {
+      disablePublicKeys.forEach((id) => {
+        if (identity.getPublicKeyById(id).getReadOnly()) {
+          result.addError(
+            new IdentityPublicKeyIsReadOnlyError(id),
+          );
+        }
+      });
+
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:64]: We need to verify getPublicKeysDisabledAt is in block time window such as createdAt in documents

File: packages/js-dpp/schema/identity/stateTransition/identityUpdate.json [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+      "minItems": 1,
+      "maxItems": 10,
+      "uniqueItems": true,
+      "items": {
+        "type": "integer",
+        "minimum": 0
+      }
+    }
+  },
+  "dependencies" : {
```
Comment (shumkov) [inline packages/js-dpp/schema/identity/stateTransition/identityUpdate.json:52]: we should use dependentRequired

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/update.ts [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+ * @param {Platform} this - bound instance class
+ * @param {Identifier|string} identityId - id of the identity to top up
+ * @param {{create: IdentityPublicKey[]; delete: IdentityPublicKey[]}} publicKeys - public keys to add
+ *
+ * @returns {boolean}
+ */
+export async function update(
+  this: Platform,
+  identityId: Identifier | string,
+  publicKeys: { create?: IdentityPublicKey[]; delete?: IdentityPublicKey[] },
```
Comment (shumkov) [inline packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/update.ts:21]: add, disable

File: packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/update.ts [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * @param {Platform} this - bound instance class
+ * @param {Identifier|string} identityId - id of the identity to top up
+ * @param {{create: IdentityPublicKey[]; delete: IdentityPublicKey[]}} publicKeys - public keys to add
+ *
+ * @returns {boolean}
+ */
+export async function update(
+  this: Platform,
+  identityId: Identifier | string,
```
Comment (shumkov) [inline packages/js-dash-sdk/src/SDK/Client/Platform/methods/identities/update.ts:20]: Usually, users will have identity already, we don't need to fetch them again

File: packages/js-dpp/lib/errors/consensus/signature/InvalidIdentityPublicKeySecurityLevelError.js [hunk 20]
```diff
@@ -0,0 +1,26 @@
+const AbstractSignatureError = require('./AbstractSignatureError');
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/signature/InvalidIdentityPublicKeySecurityLevelError.js:1]: It must be implemented in `IdentityUpdateTransition#getKeySecurityLevelRequirement`

File: packages/js-dpp/lib/errors/consensus/state/identity/IdentityPublicKeyDisabledAtWindowViolationError.js [hunk 21]
```diff
@@ -0,0 +1,44 @@
+const AbstractStateError = require('../AbstractStateError');
+
+class IdentityPublicKeyDisabledAtWindowViolationError extends AbstractStateError {
+  /**
+   * @param {Date} disabledAt
+   * @param {Date} timeWindowStart
+   * @param {Date} timeWindowEnd
+   */
+  constructor(disabledAt, timeWindowStart, timeWindowEnd) {
+    super(`Identity public key disabledAt (${disabledAt}) is out of block time window from ${timeWindowStart} and ${timeWindowEnd}`);
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/state/identity/IdentityPublicKeyDisabledAtWindowViolationError.js:10]:
```suggestion
    super(`Identity public keys disabled time (${disabledAt}) is out of block time window from ${timeWindowStart} and ${timeWindowEnd}`);
```

File: packages/js-dpp/lib/identity/IdentityFactory.js [hunk 22]
```diff
@@ -176,6 +177,38 @@ class IdentityFactory {
...
+   * Create identity update transition
+   *
+   * @param {Identity} identity - identity to update
+   * @param {{create: IdentityPublicKey[]; delete: IdentityPublicKey[]}} publicKeys - public
+   * keys to add or delete
+   * @return {IdentityUpdateTransition}
+   */
+  createIdentityUpdateTransition(
+    identity,
+    publicKeys = {},
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/IdentityFactory.js:191]: add, disable

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+      if (!result.isValid()) {
+        return result;
+      }
+
+      addPublicKeys.forEach((pk) => identity.getPublicKeys().push(pk));
+    }
+
+    result.merge(
+      validatePublicKeys(identity.getPublicKeys().map((pk) => pk.toObject())),
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:149]: I would recommend to validate disableAt here with JSON Schema instead of `validatePublicKeysAreEnabled`. You can just add an argument to `validatePublicKeys` to remove `disableAt` from schema.

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+          .forEach((pk) => {
+            if (!pk.getDisabledAt()) {
+              securityLevelsWithKeys.add(securityLevel);
+            }
+          });
+      });
+
+      SECURITY_LEVELS.forEach((securityLevel) => {
+        if (!securityLevelsWithKeys.has(securityLevel)) {
+          result.addError(new MissedSecurityLevelIdentityPublicKeyError(securityLevel));
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:89]: Shouldn't be a combination of purpose and security level? So we probably need to update error name and message and pass purpose there as well

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dpp/lib/errors/consensus/codes.js [hunk 25]
```diff
@@ -178,6 +182,10 @@ const codes = {
 
   // Identity
   4011: IdentityAlreadyExistsError,
+  4012: IdentityPublicKeyDisabledAtWindowViolationError,
+  4017: IdentityPublicKeyIsReadOnlyError,
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/codes.js:186]: Shouldn't be 13?

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/applyIdentityUpdateTransitionFactory.js [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+    await stateRepository.storeIdentity(identity);
+
+    const publicKeyHashes = identity
+      .getPublicKeys()
+      .map((publicKey) => publicKey.hash());
+
+    await stateRepository.storeIdentityPublicKeyHashes(
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/applyIdentityUpdateTransitionFactory.js:49]: You need to store only new onces.

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dpp/lib/errors/consensus/codes.js (hunk 25 above)
Comment (shuplenkov) [reply packages/js-dpp/lib/errors/consensus/codes.js:186]: 4013: InvalidDataContractVersionError,

File: packages/js-dpp/test/unit/document/stateTransition/DocumetsBatchTransition/validation/state/validateDocumentsBatchTransitionStateFactory.spec.js [hunk 27]
```diff
@@ -58,15 +58,8 @@ describe('validateDocumentsBatchTransitionStateFactory', () => {
       transitions: documentTransitions.map((t) => t.toObject()),
     }, [dataContract]);
 
-    const timeInSeconds = Math.ceil(new Date().getTime() / 1000);
-
     stateRepositoryMock = createStateRepositoryMock(this.sinonSandbox);
     stateRepositoryMock.fetchDataContract.resolves(dataContract);
-    stateRepositoryMock.fetchLatestPlatformBlockHeader.resolves({
```
Comment (antondash) [inline packages/js-dpp/test/unit/document/stateTransition/DocumetsBatchTransition/validation/state/validateDocumentsBatchTransitionStateFactory.spec.js:65]: I'm not sure why this change is here exactly? 👀

Comment (shuplenkov) [reply packages/js-dpp/test/unit/document/stateTransition/DocumetsBatchTransition/validation/state/validateDocumentsBatchTransitionStateFactory.spec.js:65]: Just mentioned that we have the same lines of code and removed the unnecessary one. You can find exactly the same code on line 72

**Review by antondash — DISMISSED** [review]
👍

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js (hunk 23 above)
Comment (shumkov) [reply packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:124]: We already validated new keys, should we validate existing keys + new keys once again?

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js [hunk 28]
```diff
@@ -38,17 +40,24 @@ function validatePublicKeysFactory(validator, bls) {
    * @typedef validatePublicKeys
    *
    * @param {RawIdentityPublicKey[]} rawPublicKeys
+   * @param {boolean} [mustBeEnabled]
    *
    * @return {ValidationResult}
    */
-  function validatePublicKeys(rawPublicKeys) {
+  function validatePublicKeys(rawPublicKeys, mustBeEnabled = false) {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:47]: We usually use options otherwise magic flags are not readable and it might create a sequence of true/false if we need to add more.

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js [hunk 29]
```diff
@@ -38,17 +40,24 @@ function validatePublicKeysFactory(validator, bls) {
...
     const result = new ValidationResult();
 
+    const clonedSchema = lodashCloneDeep(publicKeySchema);
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:50]: Let's clone only if needed

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js [hunk 30]
```diff
@@ -38,17 +40,24 @@ function validatePublicKeysFactory(validator, bls) {
...
+    if (mustBeEnabled) {
+      delete clonedSchema.properties.disabledAt;
+    }
+
     // Validate public key structure
     rawPublicKeys.forEach((rawPublicKey) => {
       result.merge(
         validator.validate(
-          publicKeySchema,
+          clonedSchema,
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:60]: The cloned schema should have a different `$id` otherwise it will be conflicts in AJV cache. Look at how we do it when we clone contracts

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+      }
+
+      const identityPublicKeys = identity.getPublicKeys();
+      addPublicKeys.forEach((pk) => identityPublicKeys.push(pk));
+      identity.setPublicKeys(identityPublicKeys);
+    }
+
+    const rawPublicKeys = identity.getPublicKeys().map((pk) => pk.toObject());
+
+    result.merge(
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:115]: I would move it up to after we apply disablePublicKeys and return if error

File: packages/js-dpp/lib/blockTimeWindow/isTimeInBlockTimeWindow.js [hunk 32]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * @param {number} lastBlockHeaderTime
+ * @param {number} timeToCheck
+ * @returns {boolean}
+ */
+function isTimeInBlockTimeWindow(lastBlockHeaderTime, timeToCheck) {
+  const {
+    timeWindowStart,
+    timeWindowEnd,
+  } = getBlockTimeWindowRange(lastBlockHeaderTime);
```
Comment (shumkov) [inline packages/js-dpp/lib/blockTimeWindow/isTimeInBlockTimeWindow.js:13]: We call this function two times. What if we rename `isTimeInBlockTimeWindow`  to `validateTimeInBlockTimeWindow` and return `ValidationResult`. In this case we can add data to validation result with time window start and end

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dpp/lib/blockTimeWindow/isTimeInBlockTimeWindow.js (hunk 32 above)
Comment (shuplenkov) [reply packages/js-dpp/lib/blockTimeWindow/isTimeInBlockTimeWindow.js:13]: But we use different errors every time

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js (hunk 28 above)
Comment (shuplenkov) [reply packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:47]: answer for the first question yes. e.g. ids might be non unique

File: packages/js-dpp/lib/blockTimeWindow/isTimeInBlockTimeWindow.js (hunk 32 above)
Comment (shumkov) [reply packages/js-dpp/lib/blockTimeWindow/isTimeInBlockTimeWindow.js:13]: Agreed to create own `ValidationResult` class with `isValid`, `getTimeWindowStart` and `getTimeWindowEnd`.

File: packages/js-dpp/lib/identity/validation/validateRequiredPurposeAndSecurityLevelFactory.js [hunk 33]
```diff
@@ -10,19 +10,19 @@ const MASTER_SECURITY_LEVEL = IdentityPublicKey.SECURITY_LEVELS.MASTER;
...
    *
    * @param {RawIdentityPublicKey[]} rawPublicKeys
    *
    * @return {ValidationResult}
    */
-  function validatePublicKeysInIdentityCreateTransition(rawPublicKeys) {
+  function validateRequiredPurposeAndSecurityLevel(rawPublicKeys) {
     const result = new ValidationResult();
 
     // Count how many purpose/security key combinations are here
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validateRequiredPurposeAndSecurityLevelFactory.js:28]: You need to check only enabled keys (filter out disabled keys before `rawPublicKeys.forEach`)

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js (hunk 31 above)
Comment (shumkov) [reply packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:115]: There is a case when we could add master and disable in one transition

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 34]
```diff
@@ -0,0 +1,133 @@
...
+    const addPublicKeys = stateTransition.getPublicKeysToAdd();
+    if (addPublicKeys) {
+      // check that all adding public keys don't contain disabledAt field
+      result.merge(
+        validatePublicKeys(addPublicKeys.map((pk) => pk.toObject()), true),
+      );
+
+      if (!result.isValid()) {
+        return result;
+      }
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:106]: We are checking this already in basic validation

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 35]
```diff
@@ -0,0 +1,133 @@
...
+    const rawPublicKeys = identity.getPublicKeys().map((pk) => pk.toObject());
+
+    result.merge(
+      validateRequiredPurposeAndSecurityLevel(rawPublicKeys),
+    );
+
+    if (!result.isValid()) {
+      return result;
+    }
+
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:122]: Let's add a comment that we are validating new fields with existing once to make sure that keys are unique and so on. We should move it under `if (addPublicKeys) {` condition

File: packages/platform-test-suite/test/functional/platform/Identity.spec.js [hunk 36]
```diff
@@ -451,7 +452,81 @@ describe('Platform', () => {
...
+        expect(identity.getRevision()).to.equal(identityBeforeUpdate.getRevision() + 1);
+        expect(identity.getPublicKeyById(2)).to.exist();
+        expect(identity.getPublicKeyById(2).getDisabledAt()).to.be.at.least(now);
+
+        expect(identity.getPublicKeyById(0)).to.exist();
+        expect(identity.getPublicKeyById(0).getDisabledAt()).to.be.undefined();
+      });
+    });
+
+    describe.skip('Masternodes', () => {
```
Comment (shumkov) [inline packages/platform-test-suite/test/functional/platform/Identity.spec.js:529]: We probably shouldn't skip it

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js (hunk 29 above)
Comment (shumkov) [reply packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:50]: And keep this cloned object in module variable to do not clone it every time

Comment (shumkov) [reply packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:50]: And keep this cloned object in module variable to do not clone it every time

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js [hunk 37]
```diff
@@ -38,17 +40,28 @@ function validatePublicKeysFactory(validator, bls) {
...
-  function validatePublicKeys(rawPublicKeys) {
+  function validatePublicKeys(rawPublicKeys, options = {}) {
     const result = new ValidationResult();
 
+    const { mustBeEnabled } = options;
+
+    let schemaToValidate = publicKeySchema;
+
+    if (mustBeEnabled) {
+      schemaToValidate = lodashCloneDeep(publicKeySchema);
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:55]: Clone only once.

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js [hunk 38]
```diff
@@ -38,17 +40,28 @@ function validatePublicKeysFactory(validator, bls) {
...
+   * @param {{mustBeEnabled: boolean}} [options]
    *
    * @return {ValidationResult}
    */
-  function validatePublicKeys(rawPublicKeys) {
+  function validatePublicKeys(rawPublicKeys, options = {}) {
     const result = new ValidationResult();
 
+    const { mustBeEnabled } = options;
+
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:51]: We need to validate that we have no more than 32 keys

File: packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js [hunk 39]
```diff
@@ -0,0 +1,132 @@
...
+    const publicKeyIdsToDisable = stateTransition.getPublicKeyIdsToDisable();
+
+    if (publicKeyIdsToDisable) {
+      publicKeyIdsToDisable.forEach((id) => {
+        if (!identity.getPublicKeyById(id)) {
+          result.addError(
+            new InvalidIdentityPublicKeyIdError(id),
+          );
+        } else if (identity.getPublicKeyById(id).getReadOnly()) {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.js:52]: `isReadOnly`

File: packages/js-dpp/test/integration/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.spec.js [hunk 40]
```diff
@@ -0,0 +1,246 @@
...
+    await stateTransition.signByPrivateKey(privateKey, IdentityPublicKey.TYPES.ECDSA_SECP256K1);
+
+    fakeTime = this.sinonSandbox.useFakeTimers(new Date());
+  });
+
+  afterEach(() => {
+    fakeTime.reset();
+  });
+
+  it('should return InvalidIdentityRevisionError', async () => {
```
Comment (shumkov) [inline packages/js-dpp/test/integration/identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory.spec.js:66]: Please make test case titles consistent

File: packages/js-dpp/test/integration/stateTransition/AbstractStateTransitionIdentitySigned.spec.js [hunk 41]
```diff
@@ -233,6 +234,19 @@ describe('AbstractStateTransitionIdentitySigned', () => {
 
       expect(isValid).to.be.true();
     });
+
+    it('should throw an error if public key is disabled', async () => {
```
Comment (shumkov) [inline packages/js-dpp/test/integration/stateTransition/AbstractStateTransitionIdentitySigned.spec.js:238]: PublicKeyIsDisabledError

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dpp/lib/errors/consensus/basic/identity/MaxIdentityPublicKeyLimitReachedError.js [hunk 42]
```diff
@@ -0,0 +1,25 @@
+const AbstractBasicError = require('../AbstractBasicError');
```
Comment (shumkov) [inline packages/js-dpp/lib/errors/consensus/basic/identity/MaxIdentityPublicKeyLimitReachedError.js:1]: We don't need this error for basic validation

File: packages/js-dpp/lib/identity/validation/validateUpdatedPublicKeys.js [hunk 43]
```diff
@@ -0,0 +1,74 @@
+const ValidationResult = require('../../validation/ValidationResult');
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validateUpdatedPublicKeys.js:1]: Move to `identity/stateTransition/IdentityUpdateTransition/validation/state`

File: packages/js-dpp/lib/identity/validation/validateUpdatedPublicKeys.js [hunk 44]
```diff
@@ -0,0 +1,74 @@
...
+    result.addError(
+      new DuplicatedIdentityPublicKeyIdError(duplicatedIds),
+    );
+  }
+
+  // Check that there's no duplicated keys
+  const keysCount = {};
+  const duplicatedKeyIds = [];
+  rawPublicKeys.forEach((rawPublicKey) => {
+    const dataHex = rawPublicKey.data.toString('hex');
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validateUpdatedPublicKeys.js:55]: Check only enabled keys

File: packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js [hunk 45]
```diff
@@ -38,17 +45,34 @@ function validatePublicKeysFactory(validator, bls) {
...
    * @param {RawIdentityPublicKey[]} rawPublicKeys
+   * @param {{mustBeEnabled: boolean}} [options]
    *
    * @return {ValidationResult}
    */
-  function validatePublicKeys(rawPublicKeys) {
+  function validatePublicKeys(rawPublicKeys, options = {}) {
     const result = new ValidationResult();
 
+    if (rawPublicKeys.length > identitySchema.properties.publicKeys.maxItems) {
```
Comment (shumkov) [inline packages/js-dpp/lib/identity/validation/validatePublicKeysFactory.js:55]: We don't need to check it here since it will be checked by the parent JSON schema (identity, identity create transition)

**Review by shumkov — CHANGES_REQUESTED** [review]

**Review by shumkov — CHANGES_REQUESTED** [review]

**Review by shumkov — APPROVED** [review]
Wow! Great job! 👍
